newapi/ooniapi/auth.py

```diff
@@ -0,0 +1,406 @@
+    smtphost = conf["MAIL_SERVER"]
+    port = conf["MAIL_PORT"]
+    mail_user = conf["MAIL_USERNAME"]
+    mail_password = conf["MAIL_PASSWORD"]
+    use_ssl = conf["MAIL_USE_SSL"]
+    log.debug(f"connecting to {smtphost}:{port} as {mail_user}")
+    SMTP = smtplib.SMTP_SSL if use_ssl else smtplib.SMTP
+    try:
+        with SMTP(host=smtphost, port=port) as s:
+            s.ehlo()
+            s.login(mail_user, mail_password)
+            s.send_message(msg)
+    except Exception as e:
+        log.error(e, exc_info=1)
+        raise
+
+
+def send_login_email(dest_addr, nick, token: str) -> None:
+    """Format and send a registration/login  email"""
+    src_addr = current_app.config["MAIL_SOURCE_ADDRESS"]
+    baseurl = current_app.config["BASE_URL"]
+    url = urljoin(baseurl, f"/api/v1/user_login?k={token}")
+
+    msg = EmailMessage()
+    msg["Subject"] = "OONI Account activation"
+    msg["From"] = src_addr
+    msg["To"] = (dest_addr,)
+
+    txt = f"""Welcome to OONI, {nick}.
+
+    Please login by following {url}
+
+    The link can be used on multiple devices and will expire in 24 hours.
+    """
+    msg.set_content(txt)
+    html = f"""\
+<html>
+  <head></head>
+  <body>
+    <p>Welcome to OONI, {nick}</p>
+    <p>
+        <a href="{url}">Please login here </a>
+    </p>
+    <p>The link can be used on multiple devices and will expire in 24 hours.</p>
+  </body>
+</html>
+"""
+    msg.add_alternative(html, subtype="html")
+    _send_email(dest_addr, msg)
+
+
+@metrics.timer("user_register")
+@auth_blueprint.route("/api/v1/user_register", methods=["POST"])
+def user_register():
+    """Auth Services: start email-based user registration
+    ---
+    parameters:
+      - in: body
+        name: register data
+        description: Registration data as HTML form or JSON
+        required: true
+        schema:
+          type: object
+          properties:
+            nickname:
+              type: string
+            email_address:
+              type: string
+    responses:
+      200:
+        description: Confirmation
+    """
+    log = current_app.logger
+    req = request.json if request.is_json else request.form
+    nick = req.get("nickname", "").strip()
+    # Accept all alphanum including unicode and whitespaces
+    if not nick.replace(" ", "").isalnum():
+        return jerror("Invalid user name")
+    if len(nick) < 3:
+        return jerror("User name is too short")
+    if len(nick) > 50:
+        return jerror("User name is too long")
+
+    email_address = req.get("email_address", "").strip().lower()
+    if not nick or not email_address:
+        return jerror("Invalid request")
+    if EMAIL_RE.fullmatch(email_address) is None:
+        return jerror("Invalid email address")
+
+    account_id = hash_email_address(email_address)
+    now = datetime.utcnow()
+    expiration = now + timedelta(days=1)
+    # On the backend side the registration is stateless
+    payload = {
+        "nbf": now,
```

Comment:
We might want to account for a small amount of "leeway" to take into account possible issues of invalid tokens due to clock skew (see: https://tools.ietf.org/html/rfc7519#section-4.1.5)
It might be enough to just remove 1 minute from `now`.

Reply:
Actually this comment is not really so much relevant for this section where we send the email. It's more relevant for the other places in which a token that needs to be used via cookies is issued.

Reply:
The timestamping and verification is done by the same host.